README.md
![PhearJS](http://d3jtdrwnfjguwh.cloudfront.net/logo-red.svg) (0.6.2)

PhearJS renders webpages. It runs a server which supervises a set number of PhantomJS workers that do the actual parsing and evaluation.

Many websites rely on AJAX and front-end rendering. When a *machine* requests a page from such a website it sees a completely different page than you would see when viewing it in a browser.

This is a problem when you want to scrape such a website or create a static copy of your dynamic site for data mining or SEO purposes. PhearJS helps you with this, by rendering pages in a headless [PhantomJS](http://phantomjs.org/) browser and returning a JSON containing the rendered page and meta data about the response.

PhearJS was built and tested on Ubuntu 15.04, but should work on most Linux distributions. OSX reportedly works fine as well.

## Example

### Request

```bash
curl -X GET "http://localhost:8100? \
  fetch_url=http%3A%2F%2Fsuch-website.com"
```

### Response

``` json
{
  "success": true,
  "input_url": "http://such-website.com",
  "final_url": "http://www.such-website.com/",
  "request_headers": {},
  "response_headers": {
    "date": "Sun, 08 Feb 2015 15:11:22 GMT",
    "content-encoding": "gzip",
    "expires": "Sun, 08 Feb 2015 15:12:33 GMT",
    "vary": "Accept-Encoding",
    "cache-control": "max-age=60",
    "last-modified": "Sun, 08 Feb 2015 15:11:33 GMT",
    "content-type": "text/html; charset=utf-8"
  },
  "had_js_errors": false,
  "content": "<rendered HTML>"
}
```

## Installation, set-up and running

Check out [these instructions](INSTALLATION.md).

## Usage

Requests can be made to:

`http://localhost:8100?fetch_url=http%3A%2F%2Fdomain.com`

PhearJS accepts the following parameters:

- **fetch_url**=<*url-encoded-string*\> ***required***
  The URL to fetch, encoded as it would be by e.g. encodeURIComponent().

- **parse_delay**=<*milliseconds*\>
  Time to wait before returning, allowing scripts and AJAX calls to run.
  Default: *as set in config.json*.

- **force**=[*false*|*true*]
  Force a cache refresh.
  Default: *false*.

- **raw**=[*false*|*true*]
  Return the raw body instead of a JSON.
  Default: *false*.

- **get_requests**=[*false*|*true*]
  In the JSON, include all HTTP requests generated by the page.
  Default: *false*.

- **get_cookies**=[*false*|*true*]
  In the JSON, include all cookies generated by the page.
  Default: *false*.

- **headers**=<*URL-encoded JSON*\>
  Add additional headers to the request.
  Default: *{}*.

- **cache_namespace**=<*string*\>
  A namespace to use on the cache. Can be useful for multi-client settings.
  Default: *global-*.
  
- **viewport_width**=<*number*\>
  Width in pixels. Default: 960
  
- **viewport_height**=<*number*\>
  height in pixels. Default: 540
  
- **as_image**=[*false*|*true*]
  Include `rendered` field to the response with the path to saved screenshot of the page.
  Format, quality, path to save the screenshot can be set in `as_image_config`.
  
  ```
        "as_image_config": {
          "format": "PNG", // PNG and JPEG are recommended among others. 
          "quality": 75, // details: http://phantomjs.org/api/webpage/method/render.html
          "path": "screenshots/"
        }
  ```
  
  Images are saved by phantomJS on day-by-day basis with some unique names as: `screenshots/2016-07-31/01:05:25.116-c5fsf.PNG`
  
  Image rendering is not perfect. Just try `http://localhost:8100/?fetch_url=http://phantomjs.org&as_image=true` 
  to see possible defects. Default: *false*.
  
### Status page

When PhearJS is running you can find a status page at `http://localhost:8100/status`. It
shows some stats about the running process. In production this status page should be enabled
via `config.json` and is password-protected through Basic Auth.

## Related projects

* [phearb](https://github.com/joaquinrulin/phearb): A Ruby gem for simplifying connection to Phear servers.
* [phearjs-express](https://github.com/Tomtomgo/phearjs-express): An Express middleware to serve prerendered pages to bots and search engines.

## Issues

If any issues may occur, please create a GitHub issue. This will help using PhearJS. Even
better: contribute! (keep reading... ;))

## Contributing

Contributions are always welcome!

In any case make sure:

1. Open an issue in this repository.
2. Fork the project.
3. Do the codes.
4. Build with `gulp build`.
5. Update these docs if applicable.
6. Make a pull request referencing the issue.

## Credits

This project was initiated at [Shuffler.fm](http://shuffler.fm) and continued at [Achieved.co](http://achieved.co).

## License

PhearJS is copyright 2014-2015 Tom Aizenberg, Marcel Gonzalez Corso.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL TOM AIZENBERG BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

### 0.6.2 (28-04-2017)

Fix headers parameter not being parsed correctly (#15, thanks @Deatheart)

### 0.6.1 (10-10-2016)

Changed Jade dependency to Pug

### 0.6.0 (10-10-2016)

Thanks to @ITsvetkoFF:

Added image rendering option

Option to add requests made by webpage to result

### 0.5.0 (04-03-2016)

Thanks to @andersju:

Option to add requests made by webpage to result
Option to add cookies created by webpage to result

### 0.4.1 (16-10-2015)

Thanks to @falhashimi for amazing feedback on PhearJS in a production environment.

Catch persistent 100% CPU consumption by PhantomJS after continuous high load
Catch rare Qt garbage collection issue

### 0.4.0 (28-09-2015)

Added a status page
Fixed favicon
Improved docs

### 0.3.1 (16-09-2015)

Added version information on startup
Workaround for [memory leak in QT](https://bugreports.qt.io/browse/QTBUG-38857) ([relevant PhantomJS issue](https://github.com/ariya/phantomjs/issues/12903))

### 0.3.0 (03-09-2015)

Added connection limiter to prevent failure due to overloading
Improved/fixed header passing to workers
Improved response on server error

### 0.2.0 (26-08-2015)

Added memcached pool
Added thread count for request handler
Code cleanup (whitespace/comments)
Fix query string parameter handling for raw/force
Improved error handling for crashed workers
Improved package.json
Improved process tree killing
Improved random worker selection

### 0.1.0 (24-03-2015)

First release

test.js

var rqs = 0;
var fs = require('fs');
var request = require('request')

var urls = []

fs.readFile('../urlstest', 'utf8', function(err, data){
  console.log(data)
  urls = data.split("\n");
  setInterval(fetch, 50);
});

function fetch(){
  var url = urls[Math.floor(Math.random()*urls.length)];
  request({url: "http://localhost:8100?force=true&fetch_url="+encodeURIComponent(url)}, function(error, response, body){
    console.log(response.statusCode, url);
  });
}

gulpfile.js
var gulp = require("gulp");
var $ = require("gulp-load-plugins")({lazy: false});
var $run = require('run-sequence');
var $logger = $.util.log;

var paths = {
  phear: ['./src/phear.coffee' ],
  scripts: ['./src/*.coffee', '!./src/phear.coffee' ],
  views: ['./src/views/*']
}

gulp.task('coffee', function(done) {

  gulp.src(paths.scripts)
    .pipe($.coffee({bare: false}).on('error', $logger))
    .pipe(gulp.dest('./lib'))
    .pipe($.size({showFiles: true}));

  gulp.src(paths.phear)
    .pipe($.coffee({bare: false}).on('error', $logger))
    .pipe($.insert.prepend("#! /usr/bin/env node\n"))
    .pipe($.chmod(755))
    .pipe(gulp.dest('./'))
    .pipe($.size({showFiles: true}));

});

gulp.task('views', function(done) {

  gulp.src(paths.views)
    .pipe(gulp.dest('./lib/views'))
    .pipe($.size({showFiles: true}));

});

gulp.task('build', function(callback) {
  $run("coffee", callback);
  $run("views", callback);
});

gulp.task('default', ['build']);


lib/stats.js
(function() {
  var Stats, dot, strftime, usage;

  dot = require('dot-object');

  strftime = require('strftime');

  usage = require('usage');

  Stats = (function() {
    function Stats(options) {
      var ref, ref1, ref2, ref3;
      this.requests = {
        active: 0,
        fail: 0,
        ok: 0,
        refuse: 0,
        total: 0
      };
      this.general = {
        start_datetime: strftime("%Y-%m-%d %H:%M:%S (%z)"),
        mode: (options != null ? (ref = options.general) != null ? ref.mode : void 0 : void 0) || void 0,
        version: (options != null ? (ref1 = options.general) != null ? ref1.version : void 0 : void 0) || void 0,
        config_file: (options != null ? (ref2 = options.general) != null ? ref2.config_file : void 0 : void 0) || void 0,
        port: (options != null ? (ref3 = options.general) != null ? ref3.port : void 0 : void 0) || void 0
      };
      if (options.config != null) {
        this.config = dot.dot(options.config);
        if (this.config['status_page.pass'] != null) {
          this.config['status_page.pass'] = "************";
        }
      } else {
        this.config = void 0;
      }
      this.workers = {};
    }

    Stats.prototype.get = function(get_worker_states, callback) {
      var i, j, len, lookups_to_complete, ref, results, worker;
      this.requests.total = this.requests.ok + this.requests.fail + this.requests.refuse;
      lookups_to_complete = this.workers.length;
      if (get_worker_states) {
        ref = this.workers;
        results = [];
        for (i = j = 0, len = ref.length; j < len; i = ++j) {
          worker = ref[i];
          results.push(((function(_this) {
            return function(worker, i) {
              return usage.lookup(worker.process.pid, function(err, res) {
                lookups_to_complete -= 1;
                if (err) {
                  _this.workers[i].usage = {
                    memory: void 0,
                    cpu: void 0
                  };
                } else {
                  _this.workers[i].usage = {
                    memory: Math.round(100 * res.memoryInfo.rss / 1024 / 1024) / 100,
                    cpu: Math.round(100 * res.cpu) / 100
                  };
                }
                if (lookups_to_complete <= 0) {
                  return callback();
                }
              });
            };
          })(this))(worker, i));
        }
        return results;
      } else {
        return callback();
      }
    };

    return Stats;

  })();

  module.exports = Stats;

}).call(this);


lib/logger.js
(function() {
  var Logger, strftime;

  strftime = require('strftime');

  Logger = (function() {
    function Logger(config, port) {
      this.config = config;
      this.port = port;
    }

    Logger.prototype.info = function(this_inst, message) {
      var id;
      id = this.space_pad_id(this_inst, 20);
      return console.log((this.date_string()) + " [" + id + "] " + message);
    };

    Logger.prototype.error = function(this_inst, message) {
      var id;
      id = this.space_pad_id(this_inst, 20);
      return console.error("\x1b[1m\x1b[31m" + (this.date_string()) + " [" + id + "] " + message + "\x1b[0m");
    };

    Logger.prototype.date_string = function() {
      return strftime("%Y-%m-%d %H:%M:%S");
    };

    Logger.prototype.space_pad_id = function(id, length) {
      id = id + ":" + this.port;
      while (id.length < length) {
        id = " " + id;
      }
      return id;
    };

    return Logger;

  })();

  module.exports = Logger;

}).call(this);


lib/worker.js
(function() {
  var Logger, close_response, config, fetch_url, ip_allowed, jar, logger, mommy, next_instance_number, page, parse_config, run_server, strip_scripts, system, url, webserver;

  run_server = function() {
    var server, service;
    server = webserver.create();
    service = server.listen(config.port, function(request, response) {
      var as_image, as_image_config, err, escaped_fetch_url, get_cookies, get_requests, parse_delay, parsed_fetch_url, ref, ref1, ref2, ref3, ref4, ref5, ref6, ref7, request_headers, request_url, this_inst, viewport_height, viewport_width;
      this_inst = next_instance_number();
      this_inst = "worker:p-" + this_inst;
      logger.info(this_inst, "Spawn subprocess.");
      if (request.headers["real-ip"] != null) {
        logger.info(this_inst, "Request origin: " + request.headers['real-ip']);
      }
      if (config.environment !== "development" && !ip_allowed(request.headers["real-ip"])) {
        response.statusCode = 403;
        return close_response(this_inst, "Forbidden.", response);
      }
      request_url = url.parse(request.url, true);
      if ((((ref = request_url.query) != null ? ref.fetch_url : void 0) == null) || !request_url.query.fetch_url) {
        response.statusCode = 400;
        return close_response(this_inst, "No URL requested.", response);
      }
      escaped_fetch_url = decodeURIComponent(request_url.query.fetch_url);
      parsed_fetch_url = url.parse(escaped_fetch_url);
      if (!("protocol" in parsed_fetch_url && ["http:", "https:"].indexOf(parsed_fetch_url.protocol) > -1)) {
        response.statusCode = 400;
        return close_response(this_inst, "Requested URL has a non-supported protocol, use http or https.", response);
      }
      parse_delay = ((ref1 = request_url.query) != null ? ref1.parse_delay : void 0) || config.parse_delay;
      get_requests = ((ref2 = request_url.query) != null ? ref2.get_requests : void 0) || config.get_requests;
      get_cookies = ((ref3 = request_url.query) != null ? ref3.get_cookies : void 0) || config.get_cookies;
      as_image = ((ref4 = (ref5 = request_url.query) != null ? ref5.as_image : void 0) === "true" || ref4 === "1") || config.as_image;
      as_image_config = config.as_image_config;
      viewport_width = ((ref6 = request_url.query) != null ? ref6.viewport_width : void 0) || config.viewport_width;
      viewport_height = ((ref7 = request_url.query) != null ? ref7.viewport_height : void 0) || config.viewport_height;
      request_headers = {};
      if (request_url.query.headers != null) {
        try {
          request_headers = JSON.parse(decodeURIComponent(request_url.query.headers));
        } catch (error) {
          response.statusCode = 400;
          return close_response(this_inst, "Malformed request headers.", response);
        }
      }
      logger.info(this_inst, "Fetching " + escaped_fetch_url + ".");
      try {
        return fetch_url(escaped_fetch_url, response, this_inst, parse_delay, request_headers, get_requests, get_cookies, as_image, as_image_config, viewport_width, viewport_height);
      } catch (error) {
        err = error;
        response.statusCode = 500;
        return close_response(this_inst, "Error on fetching.", response);
      }
    });
    if (service) {
      return logger.info("worker", "Running PhantomJS worker.");
    }
  };

  fetch_url = function(url, response, this_inst, parse_delay, request_headers, get_requests, get_cookies, as_image, as_image_config, viewport_width, viewport_height) {
    var cookie_inst, done, final_url, had_js_errors, headers, page_inst, requests;
    final_url = url;
    headers = {};
    requests = [];
    had_js_errors = false;
    done = false;
    page_inst = page.create();
    page_inst.settings.userAgent = config.user_agent;
    page_inst.settings.resourceTimeout = config.timeout;
    page_inst.customHeaders = request_headers;
    cookie_inst = jar.create();
    page_inst.cookieJar = cookie_inst;
    page_inst.onUrlChanged = function(targetUrl) {
      logger.info(this_inst, "Redirected to " + targetUrl);
      return final_url = targetUrl;
    };
    page_inst.onResourceReceived = function(response) {
      var decoded_url;
      decoded_url = decodeURIComponent(response.url);
      if (decoded_url === final_url && response.stage === "end") {
        return headers[decoded_url] = response.headers;
      }
    };
    if (get_requests === "true" || get_requests === "1") {
      page_inst.onResourceRequested = function(request) {
        return requests.push(request);
      };
    }
    page_inst.onResourceTimeout = function(response) {
      return logger.info(this_inst, "ResourceTimeout on " + url);
    };
    page_inst.onResourceError = function(msg, trace) {
      if (msg.url != null) {
        logger.info(this_inst, "ResourceError on " + url + ": " + msg.errorString + " (" + msg.url + ")");
        if (msg.url === final_url) {
          if (page_inst.parse_wait != null) {
            clearTimeout(page_inst.parse_wait);
          }
          response.statusCode = 500;
          page_inst.close();
          cookie_inst.close();
          return close_response(this_inst, "Failed to fetch this URL: " + msg.errorString, response);
        }
      } else {
        return logger.info(this_inst, "ResourceError on " + url + ": " + msg.errorString);
      }
    };
    page_inst.onError = function(msg, trace) {
      if (msg.url != null) {
        logger.info(this_inst, "JavaScriptError on " + url + ": " + msg.errorString + " (" + msg.url + ")");
      } else {
        logger.info(this_inst, "JavaScriptError on " + url + ": " + msg.errorString);
      }
      return had_js_errors = true;
    };
    return page_inst.open(url, function(status) {
      var fetch_url_headers, i;
      if (done) {
        return true;
      } else {
        done = true;
      }
      if (status !== "success") {
        logger.info(this_inst, "Failed " + url);
        response.statusCode = 500;
        close_response(this_inst, "Failed to fetch this URL.", response);
        page_inst.close();
        return cookie_inst.close();
      } else {
        logger.info(this_inst, "Fetched " + url + " parsing with a parse_delay of " + parse_delay + " ms.");
        fetch_url_headers = {};
        for (i in headers[final_url]) {
          fetch_url_headers[headers[final_url][i]["name"].toLowerCase()] = headers[final_url][i]["value"];
        }
        response.setHeader("content-type", "application/json");
        return page_inst.parse_wait = setTimeout((function() {
          var iso_date, path_to_image;
          if (!(page_inst != null ? page_inst.hasOwnProperty('content') : void 0)) {
            response.statusCode = 500;
            close_response(this_inst, "Rendering " + url + " failed.", response);
            if (page_inst != null) {
              logger.info(this_inst, "Closed page instance.");
              page_inst.close();
              cookie_inst.close();
            }
            return;
          }
          if (as_image) {
            page_inst.viewportSize = {
              width: viewport_width,
              height: viewport_height
            };
            iso_date = new Date().toISOString();
            path_to_image = "" + as_image_config.path + (iso_date.substr(0, 10)) + "/" + (iso_date.substr(11, 12)) + " - " + (Math.random().toString(36).substring(2, 7)) + "." + as_image_config.format;
            page_inst.render(path_to_image, {
              format: as_image_config.format,
              quality: as_image_config.quality
            });
          }
          response.statusCode = 200;
          response.write(JSON.stringify({
            success: true,
            input_url: url,
            final_url: final_url,
            request_headers: request_headers,
            response_headers: fetch_url_headers,
            requests: get_requests === "true" || get_requests === "1" ? requests : void 0,
            cookies: get_cookies === "true" || get_cookies === "1" ? cookie_inst.cookies : void 0,
            had_js_errors: had_js_errors,
            content: strip_scripts(page_inst.content),
            rendered: as_image ? path_to_image : void 0
          }));
          close_response(this_inst, status, response);
          page_inst.close();
          cookie_inst.close();
        }), parse_delay);
      }
    });
  };

  close_response = function(inst, status, response) {
    logger.info(inst, "Ending subprocess.");
    if ([400, 403, 500].indexOf(response.statusCode) > -1) {
      response.write(JSON.stringify({
        success: false,
        reason: status
      }));
    }
    response.close();
    return logger.info(inst, "Ended subprocess with status " + status.toUpperCase() + ".");
  };

  strip_scripts = function(doc) {
    return doc.replace(/<script(?:.*?)>(?:[\S\s]*?)<\/script>/gi, "");
  };

  next_instance_number = function() {
    return mommy.spawned_instances = mommy.spawned_instances > 10000 ? 1 : mommy.spawned_instances + 1;
  };

  ip_allowed = function(ip) {
    return config.allowed_clients.indexOf(ip) !== -1;
  };

  parse_config = function() {
    var config;
    config = new RegExp(/--config=(.*)/).exec(system.args);
    return JSON.parse(config[1]);
  };

  page = require("webpage");

  webserver = require("webserver");

  system = require("system");

  url = require("url");

  jar = require("cookiejar");

  Logger = require("./logger.js");

  config = parse_config();

  logger = new Logger(config, config.port);

  mommy = this;

  mommy.spawned_instances = 0;

  run_server();

}).call(this);


phear.js
#! /usr/bin/env node
(function() {
  var Config, Logger, Memcached, Stats, active_request_handlers, argv, basic_auth, close_response, config, do_with_random_worker, express, get_running_workers, handle_request, ip_allowed, logger, memcached, memcached_options, mode, mommy, next_thread_number, package_definition, request, respawn, serve, spawn, stats, stop, tree_kill, url, workers,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  spawn = function(n) {
    var _, i, j, len, results, worker_config;
    results = [];
    for (i = j = 0, len = workers.length; j < len; i = ++j) {
      _ = workers[i];
      workers[i] = {
        process: null,
        port: config.worker.port
      };
      worker_config = JSON.stringify(config.worker);
      workers[i].process = respawn(["phantomjs", "--disk-cache=no", "--ignore-ssl-errors=yes", "--ssl-protocol=any", "lib/worker.js", "--config=" + worker_config], {
        cwd: '.',
        sleep: 1000,
        stdio: [0, 1, 2],
        kill: 1000
      });
      workers[i].process.start();
      config.worker.port += 1;
      results.push(logger.info("phear", "Worker " + (i + 1) + " of " + n + " started."));
    }
    return results;
  };

  serve = function(port) {
    var app;
    app = express();
    app.set('view engine', 'pug');
    app.set('views', './lib/views');
    app.use(express["static"]('assets'));
    app.get('/', function(req, res) {
      var running_workers_count;
      running_workers_count = get_running_workers().length;
      if (active_request_handlers >= running_workers_count * config.worker.max_connections) {
        res.statusCode = 503;
        return close_response("phear", "Service unavailable, maximum number of allowed connections reached.", res, true);
      } else {
        return handle_request(req, res);
      }
    });
    app.get('/status/:sub?', function(req, res) {
      var allowed_routes, get_worker_states, ref, ref1, user;
      if (mode !== "development") {
        if ((ref = config.status_page) != null ? ref.enabled : void 0) {
          user = basic_auth(req);
          if ((user != null ? user.pass : void 0) !== config.status_page.pass || (user != null ? user.name : void 0) !== config.status_page.name) {
            res.statusCode = 401;
            res.header('WWW-Authenticate', 'Basic realm=\nUsername and password please.');
            return res.end();
          }
        } else {
          res.statusCode = 403;
          return close_response("phear", "Forbidden.", res, true);
        }
      }
      stats.requests.active = active_request_handlers;
      stats.workers = workers;
      allowed_routes = ['general', 'workers', 'config'];
      if (ref1 = req.params.sub, indexOf.call(allowed_routes, ref1) < 0) {
        req.params.sub = 'general';
      }
      get_worker_states = req.params.sub === "workers";
      return stats.get(get_worker_states, function() {
        return res.render(req.params.sub + ".pug", {
          stats: stats,
          sub: req.params.sub
        });
      });
    });
    app.listen(port);
    return logger.info("phear", "Phear started.");
  };

  handle_request = function(req, res) {
    var cache_key, cache_namespace, respond, thread_number;
    thread_number = next_thread_number();
    res.header("Access-Control-Allow-Origin", "*");
    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
    if (mode !== "development" && !ip_allowed(req.headers["real-ip"])) {
      res.statusCode = 403;
      return close_response("phear-" + thread_number, "Forbidden.", res);
    }
    if (req.query.fetch_url == null) {
      res.statusCode = 400;
      return close_response("phear-" + thread_number, "No URL requested, you have to set fetch_url=encoded_url.", res);
    }
    if (req.query.headers != null) {
      try {
        JSON.parse(req.query.headers);
      } catch (error1) {
        res.statusCode = 400;
        return close_response("phear-" + thread_number, "Additional headers not properly formatted, e.g.: encodeURIComponent('{extra: \"Yes.\"}').", res);
      }
    }
    respond = function(statusCode, body) {
      var parsed_body, ref;
      if ((ref = req.query.raw) === "true" || ref === "1") {
        parsed_body = JSON.parse(body);
        
        if(config.worker.as_image_config.base64)
        {
          var img = new Buffer(parsed_body.rendered, "base64");
          res.header("content-type", 'image/'+config.worker.as_image_config.format);
          res.header("content-Length", img.length);
          res.send(img); 
        }else
          res.status(statusCode).send(parsed_body.content);
      } else {
        res.set("content-type", "application/json");
        res.status(statusCode).send(body);
      }
      res.end();
      stats.requests.ok += 1;
      return active_request_handlers -= 1;
    };
    active_request_handlers += 1;
    cache_namespace = "global-";
    if (req.query.cache_namespace != null) {
      cache_namespace = req.query.cache_namespace;
    }
    cache_key = "" + cache_namespace + req.query.fetch_url;
    return memcached.get(cache_key, function(error, data) {
      var ref;
      if ((error != null) || (data == null) || ((ref = req.query.force) === "true" || ref === "1")) {
        return do_with_random_worker(thread_number, function(worker) {
          var options, worker_request_url;
          worker_request_url = url.format({
            protocol: "http",
            hostname: "localhost",
            port: worker.port,
            query: req.query
          });
          options = {
            url: worker_request_url,
            headers: {
              'real-ip': req.headers['real-ip']
            },
            timeout: config.global_timeout
          };
          return request(options, function(error, response, body) {
            var err, ref1;
            try {
              if (response.statusCode === 200) {
                memcached.set(cache_key, body, config.cache_ttl, function() {
                  return logger.info("phear-" + thread_number, "Stored " + req.query.fetch_url + " in cache");
                });
              }
              return respond(response.statusCode, body);
            } catch (error1) {
              err = error1;
              res.statusCode = 500;
              close_response("phear-" + thread_number, "Request failed due to an internal server error.", res);
              if ((ref1 = worker.process.status) !== "stopping" && ref1 !== "stopped") {
                logger.info("phear-" + thread_number, "Trying to restart worker with PID " + worker.process.pid + "...");
                worker.process.stop(function() {
                  if (worker.process.status === "stopped") {
                    worker.process.start();
                    return logger.info("phear-" + thread_number, "Restarted worker with PID " + worker.process.pid + ".");
                  }
                });
              } else {
                logger.info("phear-" + thread_number, "Worker with PID " + worker.process.pid + " is being restarted...");
              }
              return active_request_handlers -= 1;
            }
          });
        });
      } else {
        logger.info("phear-" + thread_number, "Serving entry from cache.");
        return respond(200, data);
      }
    });
  };

  do_with_random_worker = function(thread_number, callback) {
    var running_workers;
    running_workers = get_running_workers();
    if (running_workers.length > 0) {
      return callback(running_workers[Math.floor(Math.random() * running_workers.length)]);
    } else {
      logger.info("phear-" + thread_number, "No running workers, waiting for a new worker to come up.");
      return setTimeout((function() {
        return do_with_random_worker(thread_number, callback);
      }), 500);
    }
  };

  get_running_workers = function() {
    var j, len, results, worker;
    results = [];
    for (j = 0, len = workers.length; j < len; j++) {
      worker = workers[j];
      if (worker.process.status === "running") {
        results.push(worker);
      }
    }
    return results;
  };

  close_response = function(inst, status, response, refused) {
    if (refused == null) {
      refused = false;
    }
    response.set("content-type", "application/json");
    logger.info(inst, "Ending process.");
    if ([400, 403, 500, 503].indexOf(response.statusCode) > -1) {
      response.status(response.statusCode).send(JSON.stringify({
        success: false,
        reason: status
      }));
    }
    response.end();
    if (refused) {
      stats.requests.refuse += 1;
    } else {
      stats.requests.fail += 1;
    }
    return logger.info(inst, "Ended process with status " + (status.toUpperCase()) + ".");
  };

  next_thread_number = function() {
    return mommy.handler_thread_number = mommy.handler_thread_number > 10000 ? 1 : mommy.handler_thread_number + 1;
  };

  ip_allowed = function(ip) {
    return config.worker.allowed_clients.indexOf(ip) !== -1;
  };

  stop = function() {
    logger.info("phear", "Trying to kill process and " + workers.length + " workers gently...");
    return tree_kill(process.pid, 'SIGTERM', function() {
      logger.info("phear", "Trying to kill process and workers forcefully...");
      return tree_kill(process.pid, 'SIGKILL');
    });
  };

  basic_auth = require('basic-auth');

  express = require('express');

  Memcached = require('memcached');

  package_definition = require('./package.json');

  request = require('request');

  respawn = require('respawn');

  tree_kill = require('tree-kill');

  url = require('url');

  argv = require('yargs').usage('Parse dynamic webpages.\nUsage: $0').example('$0 -c', 'location of phear configuration file').alias('c', 'config').example('$0 -e', 'environment to run in.').alias('e', 'environment')["default"]({
    c: "./config/config.json",
    e: "development"
  }).argv;

  Logger = require("./lib/logger.js");

  Config = require("./lib/config.js");

  Stats = require("./lib/stats.js");

  mode = argv.e;

  config = new Config(argv.c, mode).config;

  config.worker.environment = mode;

  logger = new Logger(config, config.base_port);

  workers = new Array(config.workers);

  memcached_options = config.memcached.options;

  memcached_options.poolSize = config.workers * 10;

  memcached = new Memcached(config.memcached.servers, memcached_options);

  memcached.on('issue', function(f) {
    logger.info("phear", "Memcache failed: " + f.messages);
    return stop();
  });

  memcached.stats(function(_) {
    return true;
  });

  process.on('uncaughtException', function(err) {
    logger.error("phear", "UNCAUGHT ERROR: " + err.stack);
    return stop();
  });

  logger.info("phear", "Starting Phear...");

  logger.info("phear", "==================================");

  logger.info("phear", "Version: " + package_definition.version);

  logger.info("phear", "Mode: " + mode);

  logger.info("phear", "Config file: " + argv.c);

  logger.info("phear", "Port: " + config.base_port);

  logger.info("phear", "Workers: " + config.workers);

  logger.info("phear", "==================================");

  mommy = this;

  mommy.handler_thread_number = 0;

  active_request_handlers = 0;

  stats = new Stats({
    general: {
      mode: mode,
      version: package_definition.version,
      config_file: argv.c,
      port: config.base_port
    },
    config: config
  });

  spawn(config.workers);

  serve(config.base_port);

}).call(this);
